src/haz3lcore/zipper/action/Insert.re

```diff
@@ -1,11 +1,90 @@
+  /* Removes the d-neighboring tile and reconstructs it, triggering
+     keyword-expansion; precondition: the d-neighbor should be a monotile
+     string-matching a keyword of an expanding form */
+  let (new_label, backpack) = Molds.delayed_expansion(t);
+  let+ z = delete(caret, z);
+  construct(~backpack, ~caret, new_label, z, id_gen);
+};
+
+let expand_or_barf_left_neighbor = ((z, _) as s: state): option(state) =>
+  /* If left neighbor is a monotile (a) string-matching the shard at the
+     top of the backpack, barf it, or (b) an expansing keyword, expand it. */
+  switch (left_neighbor_monotile(z.relatives.siblings)) {
+  | Some(t) when Backpack.will_barf(t, z.backpack) => barf(Left, s)
+  | Some(t) when Molds.is_delayed(t) => delayed_expand(t, Left, s)
+  | _ => Some(s)
+  };
+
+let expand_or_barf_right_neighbor = ((z, _) as s: state): option(state) =>
+  /* If right neighbor is a monotile (a) string-matching the shard at the
+     top of the backpack, barf it, or (b) an expansing keyword, expand it. */
+  switch (right_neighbor_monotile(z.relatives.siblings)) {
+  | Some(t) when Backpack.will_barf(t, z.backpack) => barf(Right, s)
+  | Some(t) when Molds.is_delayed(t) => delayed_expand(t, Right, s)
+  | _ => Some(s)
+  };
+
+let make_new_tile = (t: Token.t, caret: Direction.t, z: t): IdGen.t(t) =>
+  /* Adds a new tile at the caret. If the new token matches the top
+     of the backpack, the backpack shard is dropped. Otherwise, we
+     construct a new tile, which may immediately expand. */
+  switch (put_down(Left, z)) {
+  | Some(z') when Molds.is_instant(t) && Backpack.will_barf(t, z.backpack) =>
+    IdGen.return(z')
+  | _ =>
+    let (lbl, backpack) = Molds.instant_expansion(t);
+    construct(~caret, ~backpack, lbl, z);
+  };
+
+let expand_neighbors_and_make_new_tile =
+    (char: Token.t, state: state): option(state) => {
+  /* Trigger a token boundary event and create a new tile.
+     This process potentially involves both neighboring tiles,
+     potentially triggering up to 3 expansions or backpack barfs.
+     In particular, both left and right neighboring monotiles may
+     undergo delayed (aka keyword) expansion, and the newly-created
+     single-character token may undergo instant expansion. Currently
+     made the decision to expand or barf the neighbors before making
+     the new tile because barfing is limited to the top of the backpack,
+     and I wanted things like "if|then", when you enter a "(", to
+     barf the "then", before it is buried by the ")" added to the BP.
+     The order here could be revisited if barfing was more sophisticated.
+     */
+  let* (z, id_gen) = expand_or_barf_left_neighbor(state);
+  let (z, id_gen) = regrout(Left, z, id_gen);
+  /* Note to david: I'm not sure why the above regrout is necessary.
+     Without it, there is a Nonconvex segment error thrown in exactly
+     one case, the double barf case: insert space on "if then|else" */
```

Comment:
not sure either... my first thought was that this would be due to some new bidelimited container being created that the final regrouting doesn't reach into (eg newly created if-then container, or maybe the segment on right side of else), but I'm not sure how this regrout operation at this stage would affect that

Reply:
Ended up undoing this for now... see issue #953